### ISSUE TITLE
torch.triu() may returns wrong values using MPS

### ISSUE THREAD

**Issue by TheCluster** [pytorch/pytorch#100005]
### 🐛 Describe the bug

Using MPS, torch.triu() may returns tensor with incorrect values.

It works as expected on CPU:

```
>>> import torch
>>> 
>>> mask = torch.full((1, 1, 10, 10), float("-inf"), device=torch.device("cpu"))
>>> result = torch.triu(mask, diagonal=1)
>>> print(result)
tensor([[[[0., -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf],
          [0., 0., -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf],
          [0., 0., 0., -inf, -inf, -inf, -inf, -inf, -inf, -inf],
          [0., 0., 0., 0., -inf, -inf, -inf, -inf, -inf, -inf],
          [0., 0., 0., 0., 0., -inf, -inf, -inf, -inf, -inf],
          [0., 0., 0., 0., 0., 0., -inf, -inf, -inf, -inf],
          [0., 0., 0., 0., 0., 0., 0., -inf, -inf, -inf],
          [0., 0., 0., 0., 0., 0., 0., 0., -inf, -inf],
          [0., 0., 0., 0., 0., 0., 0., 0., 0., -inf],
          [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]]]])
```

But on MPS, it return tensor with NaN:

```
>>> import torch
>>> 
>>> mask = torch.full((1, 1, 10, 10), float("-inf"), device=torch.device("mps"))
>>> result = torch.triu(mask, diagonal=1)
>>> print(result)
tensor([[[[nan, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf],
          [nan, nan, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf],
          [nan, nan, nan, -inf, -inf, -inf, -inf, -inf, -inf, -inf],
          [nan, nan, nan, nan, -inf, -inf, -inf, -inf, -inf, -inf],
          [nan, nan, nan, nan, nan, -inf, -inf, -inf, -inf, -inf],
          [nan, nan, nan, nan, nan, nan, -inf, -inf, -inf, -inf],
          [nan, nan, nan, nan, nan, nan, nan, -inf, -inf, -inf],
          [nan, nan, nan, nan, nan, nan, nan, nan, -inf, -inf],
          [nan, nan, nan, nan, nan, nan, nan, nan, nan, -inf],
          [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]]]],
       device='mps:0')
```

Also, it works correct on CUDA:

```
>>> import torch
>>> 
>>> mask = torch.full((1, 1, 10, 10), float("-inf"), device=torch.device("cuda"))
>>> result = torch.triu(mask, diagonal=1)
>>> print(result)
tensor([[[[0., -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf],
          [0., 0., -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf],
          [0., 0., 0., -inf, -inf, -inf, -inf, -inf, -inf, -inf],
          [0., 0., 0., 0., -inf, -inf, -inf, -inf, -inf, -inf],
          [0., 0., 0., 0., 0., -inf, -inf, -inf, -inf, -inf],
          [0., 0., 0., 0., 0., 0., -inf, -inf, -inf, -inf],
          [0., 0., 0., 0., 0., 0., 0., -inf, -inf, -inf],
          [0., 0., 0., 0., 0., 0., 0., 0., -inf, -inf],
          [0., 0., 0., 0., 0., 0., 0., 0., 0., -inf],
          [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]]]], device='cuda:0')
```

This behavior is reproducible on both pytorch 2.0 and pytorch 1.12.x

### Versions

PyTorch version: 2.0.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.3.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: version 3.26.3
Libc version: N/A

Python version: 3.10.10 | packaged by conda-forge | (main, Mar 24 2023, 20:12:31) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.3.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Ultra

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] torch==2.0.0
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] pytorch                   2.0.0                  py3.10_0    pytorch

cc @ezyang @gchanan @zou3519 @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by malfet** [comment]
Added `infs-and-nans`, imo it's not a high-pri issue, but let's discuss

**Comment by TheCluster** [comment]
Yes, it is probably not a high-priority issue. But, for example, torch.triu() is used in a LLaMA model (in the original python code from Meta).

**Comment by dpvasani** [comment]
When executing torch.triu() on the CPU device, the function correctly generates the upper triangular matrix with the specified diagonal, as demonstrated by the following example:

python
Copy code
import torch

mask = torch.full((1, 1, 10, 10), float("-inf"), device=torch.device("cpu"))
result = torch.triu(mask, diagonal=1)
print(result)
Output:

css
Copy code
tensor([[[[0., -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf],
          [0., 0., -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf],
          [0., 0., 0., -inf, -inf, -inf, -inf, -inf, -inf, -inf],
          [0., 0., 0., 0., -inf, -inf, -inf, -inf, -inf, -inf],
          [0., 0., 0., 0., 0., -inf, -inf, -inf, -inf, -inf],
          [0., 0., 0., 0., 0., 0., -inf, -inf, -inf, -inf],
          [0., 0., 0., 0., 0., 0., 0., -inf, -inf, -inf],
          [0., 0., 0., 0., 0., 0., 0., 0., -inf, -inf],
          [0., 0., 0., 0., 0., 0., 0., 0., 0., -inf],
          [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]]]])
However, when executing the same code on the MPS device, the result contains NaN values instead of the expected values:

python
Copy code
import torch

mask = torch.full((1, 1, 10, 10), float("-inf"), device=torch.device("mps"))
result = torch.triu(mask, diagonal=1)
print(result)
Output:

css
Copy code
tensor([[[[nan, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf],
          [nan, nan, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf],
          [nan, nan, nan, -inf, -inf, -inf, -inf, -inf, -inf, -inf],
          [nan, nan, nan, nan, -inf, -inf, -inf, -inf, -inf, -inf],
          [nan, nan, nan, nan, nan, -inf, -inf, -inf, -inf, -inf],
          [nan, nan, nan, nan, nan, nan, -inf, -inf, -inf, -inf],
          [nan, nan, nan, nan, nan, nan, nan, -inf, -

**Comment by zou3519** [comment]
We should check if this still reproduces on nightly.

**Comment by h-sellak** [comment]
This behaviour is still reproducible on Pytorch 2.1.2.

**Comment by alex0dd** [comment]
Still reproducible in PyTorch 2.3.1

**Comment by hvaara** [comment]
Fixed in #128575.

Should we expand the test case to include the example given by OP?

xref #128429

**Comment by skotapati** [comment]
Hi @hvaara, I think this case is mostly covered by the test here: https://github.com/pytorch/pytorch/blob/9e0ddc0e14c26b22780089ba8c36854369c6d348/test/test_mps.py#L1603. However let me know if you think we should expand on it, thanks!
